### ISSUE TITLE
More generic FFT

### ISSUE THREAD

**Issue by daanhb** [MikaelSlevinsky/FastTransforms.jl#62]
The FFT provided in FastTransforms is specific to `BigFloat`. It would be nice to support other abstract float types. Any reason why `AbstractFloats` is not defined like this:
```julia
const AbstractFloats = Union{AbstractFloat,Complex{<:AbstractFloat}}
```
?

I ran a test using the `DoubleFloats.jl` package. There are a few places where `BigFloat` is hardcoded in `genericfft.jl`, but apart from that the only issues I encountered were associated either with `DoubleFloats` or Julia Base (see https://github.com/JuliaMath/DoubleFloats.jl/issues/65).

**Comment by daanhb** [comment]
I'll make a pull request and then we can see.

**Comment by MikaelSlevinsky** [comment]
Related: #34. Originally, I thought that someone from FFTW with connections to Julia might write native generic FFTs, but I suppose we can't wait forever.

**Comment by daanhb** [comment]
Oops, I did not see the existing issue.

The generic FFT code works with other AbstractFloat types, but making them accessible for any AbstractFloat like I suggested above is not free of issues. For example, since `dct` is defined for Vectors in FastTransforms but for AbstractArrays (with any dimension) in FFTW, the definition of FastTransforms.jl is actually more specific. Hence, it ends up getting called, even when the argument has Float64 elements. Furthermore, the test suite yields StackOverflowErrors, presumably for a similar reason. For the time being, the code is here: https://github.com/daanhb/FastTransforms.jl/blob/genericfft/src/fftBigFloat.jl

**Comment by daanhb** [comment]
The code now works and all tests pass, with some changes to the codeflow. I've renamed fft to generic_fft. The fft routines in AbstractFFTs.jl invoke plan_fft (and friends). We implement plan_fft with the same signature as in FFTW, but such that the methods of FFTW are more specific than ours when T is one of the fftwNumber types. We provide a fallback for any AbstractFloat. The dummy plans of FastTransforms subsequently call generic_fft (and friends).

So: fft (AbstractFFTs) -> plan_fft (FastTransforms) -> DummyFFTPlan.
The DummyFFTPlan calls `generic_fft`.

For Float64: fft (AbstractFFTs) -> plan_fft (FFTW) -> SomeFFTWPlan, much more efficient.

I've had to change the dummy plan for irfft to include the extra integer argument, which is the length of the original vector to which rfft was applied. I'll make a pull request.
